pex/auth.py

```diff
@@ -0,0 +1,93 @@
+# Copyright 2022 Pants project contributors (see CONTRIBUTORS.md).
+# Licensed under the Apache License, Version 2.0 (see LICENSE).
+
+from __future__ import absolute_import
+
+import os
+from netrc import netrc
+
+from pex.compatibility import urlparse
+from pex.typing import TYPE_CHECKING
+
+if TYPE_CHECKING:
+    from typing import Iterable, Iterator, Optional, Text, Tuple
+
+    import attr  # vendor:skip
+else:
+    from pex.third_party import attr
+
+
+@attr.s(frozen=True)
+class PasswordEntry(object):
+    @staticmethod
+    def _reduced_url(url_info):
+        # type: (urlparse.ParseResult) -> str
+        return "{scheme}://{host}{port}".format(
+            scheme=url_info.scheme,
+            host=url_info.hostname,
+            port=":{port}".format(port=url_info.port) if url_info.port is not None else "",
+        )
+
+    @classmethod
+    def maybe_extract_from_url(cls, url):
+        # type: (str) -> Optional[PasswordEntry]
+        url_info = urlparse.urlparse(url)
+        if not url_info.username or not url_info.password:
+            return None
+
+        return cls(
+            uri=cls._reduced_url(url_info),
+            username=url_info.username,
+            password=url_info.password,
+        )
+
+    username = attr.ib()  # type: str
+    password = attr.ib()  # type: str
+    uri = attr.ib(default=None)  # type: Optional[str]
+
+    def uri_or_default(self, url):
+        # type: (Text) -> str
+        return self.uri or self._reduced_url(urlparse.urlparse(url))
+
+
+@attr.s(frozen=True)
+class PasswordDatabase(object):
+    @classmethod
+    def from_netrc(cls, netrc_file="~/.netrc"):
+        # type: (Optional[str]) -> PasswordDatabase
+        if not netrc_file:
+            return cls()
+
+        netrc_path = os.path.expanduser("~/.netrc")
+        if not os.path.isfile(netrc_path):
+            return cls()
+
+        def iter_entries():
+            # type: () -> Iterator[PasswordEntry]
+            for machine, (login, _, password) in netrc(netrc_path).hosts.items():
+                if password is None:
+                    continue
+
+                if machine == "default":
+                    # The 'default' entry is special and means just that; so we omit any qualifying
+                    # URI.
+                    yield PasswordEntry(username=login, password=password)
```

Comment:
I get that this is part of how netrc works and so we should support it, but it seems bonkers to me to allow sending a password to every host. But I guess if that's how someone has configured their netrc, so be it.

Reply:
Yeah, I'm not fighting how this ancient thing works - and if I don't work how it works I only break users who have made a choice.